File: 2023/typescript/src/day1/day1.ts
import { readAsStringArray } from "../utils/getInput";

// cover cases like oneight eighthree etc...
const numberMap = new Map<string, string>([
  ["one", "o1e"],
  ["two", "t2o"],
  ["three", "t3e"],
  ["four", "f4r"],
  ["five", "f5e"],
  ["six", "s6x"],
  ["seven", "s7n"],
  ["eight", "e8t"],
  ["nine", "n9e"],
]);

const getCalibrationValue = (
  input: string[],
  useDigit: boolean = false,
): number => {
  if (useDigit) {
    for (let i = 0; i < input.length; i++) {
      numberMap.forEach(
        (val, key) => (input[i] = input[i].replaceAll(key, val)),
      );
    }
  }

  console.log(input);
  return input
    .map((line) => line.split("").filter((c) => parseInt(c)))
    .map((val) => parseInt(val[0] + val.at(-1)))
    .reduce((acc, val) => acc + val);
};

export const partOne = (filename: string): number => {
  const input = readAsStringArray(filename);
  return getCalibrationValue(input);
};

export const partTwo = (filename: string) => {
  const input = readAsStringArray(filename);
  return getCalibrationValue(input, true);
};

File: golang/2023/day1/day1.go
package day1

import (
	"aoc/golang/utils"
	"fmt"
	"strconv"
	"strings"
	"unicode"
)

func mapTextDigit(input []string) {
	numberMap := map[string]string{
		"one":   "o1e",
		"two":   "t2o",
		"three": "t3e",
		"four":  "f4r",
		"five":  "f5e",
		"six":   "s6x",
		"seven": "s7n",
		"eight": "e8t",
		"nine":  "n9e",
	}
	for i, _ := range input {
		for k, v := range numberMap {
			input[i] = strings.ReplaceAll(input[i], k, v)
		}
	}
}

func getCalibration(input []string) int {
	calibration := 0
	for _, s := range input {
		val := ""
		for _, c := range s {
			if unicode.IsDigit(c) {
				val += string(c)
			}
		}
		num, _ := strconv.Atoi(string(val[0]) + string(val[len(val)-1]))
		calibration += num
	}
	return calibration
}

func runPartOne(filename string) int {
	fmt.Println("Part One")
	input := utils.ReadAsSlice[string](filename)
	return getCalibration(input)
}

func runPartTwo(filename string) int {
	fmt.Println("Part Two")
	input := utils.ReadAsSlice[string](filename)
	mapTextDigit(input)
	return getCalibration(input)
}
